File: uporto/contest01/g.py
c = int(input())

f = int(input())
fitems = list(map(int, input().split()))

s = int(input())
sitems = list(map(int, input().split()))

def judgeNBoxes(nboxes):
    boxes = [0 for i in range(nboxes)]

    fbox = 0
    fremainingitems = f
    fitemindex = 0
    
    sbox = nboxes-1
    sremainingitems = s
    sitemindex = 0
    while(fremainingitems > 0 and sremainingitems > 0):
        if boxes[fbox] + fitems[fitemindex] <= c:
            fremainingitems -= 1
            fitemindex += 1
        else:
            fbox += 1

        if boxes[sbox] + sitems[sitemindex] <= c:
            sremainingitems -= 1
            sitemindex += 1
        else:
            sbox -= 1

    while(fbox < nboxes and fremainingitems > 0):
        if boxes[fbox] + fitems[fitemindex] <= c:
            fremainingitems -= 1
            fitemindex += 1
        else:
            fbox += 1

    while(sremainingitems > 0 and sbox >= 0):
        if boxes[sbox] + sitems[sitemindex] <= c:
            sremainingitems -= 1
            sitemindex += 1
        else:
            sbox -= 1
    
    if fremainingitems > 0 or sremainingitems > 0:
        return False
    else:
        return True

i = 0
j = f+s
answer = 0
while (i <= j):
    mid = (i+j)/2
    mid = int(mid)
    if judgeNBoxes(mid):
        answer = mid
        j = mid-1
    else:
        i = mid+1

print(answer)

File: uporto/contest01/g.cpp
#include <bits/stdc++.h>
#define ll long long
using namespace std;

ll c;
ll f;
vector<ll> fitems;
ll s;
vector<ll> sitems;

bool calcBoxes(ll nboxes){
    vector<ll> boxes(nboxes);
    ll i = f;
    ll j = s;
    ll fbox = 0;
    ll sbox = nboxes-1;
    while(i || j){
        if (i >= 0) {
            if (boxes[fbox] + fitems[f-i] <= c){
                fbox++;
                i--;
            } else if (boxes[++fbox] + fitems[f-i] <= c) {
                i--;        
            } else {
                break;
            }
        }
        if (j >= 0) {
            if (boxes[sbox] + sitems[s-j] <= c){
                sbox--;
            } else if (boxes[--sbox] + sitems[s-j] <= c) {
                j--;    
            } else {
                break;
            }
        }
    }
    if (i || j) {
        return false;
    } else {
        return true;
    }
}

int main(){
    cin >> c;
    cin >> f;
    fitems.resize(f);
    for (ll i=0; i < f; i++) cin >> fitems[i];
    cin >> s;
    for (ll i=0; i < f; i++) cin >> sitems[i]; 
    
    ll l = 0;
    ll r = f+s;
    ll answer=r;
    while (l<r) {
        ll mid = l+(r-l)/2;
        if (calcBoxes(mid)) {
            r = mid-1;
            answer = mid;
        } else {
            l = mid+1;
        }
    }

    return 0;
}
